packages/opencensus-core/src/trace/instrumentation/base-plugin.ts

```diff
@@ -60,10 +78,59 @@ export abstract class BasePlugin implements types.Plugin {
   // while unpatching is not. It should be the other way around
 
   // tslint:disable:no-any
-  applyPatch(moduleExports: any, tracer: modelTypes.Tracer, version: string):
-      any {
-    this.setPluginContext(moduleExports, tracer, version);
+  applyPatch(
+      moduleExports: any, tracer: modelTypes.Tracer, version: string,
+      basedir?: string): any {
```

Comment:
`basedir` is optional here but is passed to functions that require it to have a value later. I would suggest making it required here as well.

Reply:
It will not be used by all implementations, just the ones that need to patch internal files that are not exported. The parameter as optional does not affect current implementations.

Reply:
TS optionality doesn't refer to whether it is required to be used, but whether it is required to be given a value. In the latter case it is not optional because some plugins just won't work if you don't pass it.

I understand that you'd have to update the plugins to add a `basedir` argument -- if you agree that it should be required then it should be done in this PR.

Reply:
Method loadPlugins from PluginLoader is always passing this parameter value.
```
 result = plugin.applyPatch(exports, this.tracer, version, basedir);
```

Reply:
The implementations that will require this parameter will have it available.

Reply:
I think you might be confounding optional assignment vs. optional usage:

__optional assignment__
```js
// it is OK for us not to pass a value for `basedir`.
plugin.applyPatch(exports, this.tracer, version);
```
__optional usage__
```js
class MyPlugin extends BasePlugin {
  applyPatch(exports, tracer, version, basedir) {
    // it is OK for us not to use `basedir`.
  }
}
```

The syntax `basedir?` suggests that `basedir`'s _assignment_ is optional. But as we have both said, the real intent that we want to convey is that its _usage_ is optional. (There is no syntax for this; the only way to show this intent is to document it.)